### ISSUE TITLE
Hide internal details

### ISSUE THREAD

**Issue by emersion** [swaywm/wlroots#1828]
One thing that really pushes back stabilization for me is that all our struct fields are public. This makes sense in a lot of cases because we want compositors to have a lot of freedom. However exposing too many fields could paint us in an API compatibility corner. In particular, `wl_listener` fields shouldn't be exposed IMHO.

Additionally, our internal fields clobber the header files and make it difficult for compositor writers to know what fields they should use (example: xdg-shell geometry fields).

**Comment by ddevault** [comment]
I agree, but would also posit that simply having some details available via structs is not necessarily a bad thing, even if they're read-only and writing them would be considered a programming error. I agree that our internal `wl_listeners` are a good candidate for privatization.

One thing I don't want to do is start guaranteeing ABI stability, ever.

**Comment by emersion** [comment]
>I agree, but would also posit that simply having some details available via structs is not necessarily a bad thing, even if they're read-only and writing them would be considered a programming error.

Yes. I'm not advocating for adding a bunch of getters and setters everywhere. :P

>Would we just apply libweston's strategy - hide all currently unused APIs and then expose them again if needed in the future?

Yeah. In case of doubt, it's always easier to make a field private, and then decide to make it public later. Doing it the other way around is an API breakage.

**Comment by emersion** [comment]
Some other possible (but not necessarily good or bad) ideas:

- Prefix our private fields with `_` (suggested by ericonr)
- Have a macro hack like this:

```c
#ifdef WLR_SHOW_PRIVATE
#define WLR_PRIVATE_BEGIN
#define WLR_PRIVATE_END
#else
#define WLR_PRIVATE_BEGIN char private[sizeof(struct {
#define WLR_PRIVATE_END }];
#endif
```
